### PR TITLE
Porting WML tags to Lua

### PR DESCRIPTION
**Pull request by CelticMinstrel**
This is a branch dedicated to porting WML tags to Lua, mainly focusing on ones that are nominally implemented in Lua but actually just defer to an API call that does all of the work.

It's not ready for merge yet, but could use a lot of testing for each affected tag to ensure that things still work as they should. Since the first two commits touch AI stuff to a great extent, @mattsc should probably take a look when he has time.

Things to be tested:
* [ ] modify_side shroud/fog ops
* [ ] modify_side general ops
* [X] modify_side AI ops ([ai], switch_ai)
* [X] modify_ai
* [ ] heal_unit
* [ ] animate_unit

Other tags to port:
* [ ] kill
* [ ] shroud
* [ ] scroll

Tags to evaluate:
* [ ] music
* [ ] set_menu_item
* [ ] time_area
* [ ] replace_schedule

### CODE REVIEW

**Comment by gfgtdf** [comment]
there is no reason to keep `wesnoth.modify_side` compability since it  was added in wesnoth 1.13 and isnt in the wiki.

**Comment by CelticMinstrel** [comment]
Fair enough, I can remove that easily.

File: src/scripting/game_lua_kernel.cpp [hunk 1]
```diff
@@ -2443,15 +2595,22 @@ int game_lua_kernel::intf_play_sound(lua_State *L)
...
 int game_lua_kernel::intf_scroll_to_tile(lua_State *L)
 {
 	map_location loc = luaW_checklocation(L, 1);
 	bool check_fogged = luaW_toboolean(L, 2);
-	bool immediate = luaW_toboolean(L, 3);
+	game_display::SCROLL_TYPE scroll = luaW_toboolean(L, 4)
+		? luaW_toboolean(L, 3)
+			? game_display::WARP
+			: game_display::SCROLL
+		: luaW_toboolean(L, 3)
```
Comment (gfgtdf) [inline src/scripting/game_lua_kernel.cpp]: this looks wrong, you are cheking the third parameter twice here.

Comment (gfgtdf) [inline src/scripting/game_lua_kernel.cpp]: hmm neverming, just saw that there is a luaW_toboolean(L, 4) above.

Comment (gfgtdf) [inline src/scripting/game_lua_kernel.cpp]: ok but then the releasenotes is likeley wrong, since it now accepts a fourth argument, not a third

**Comment by gfgtdf** [comment]
same is true for wesnoth.animate_unit, wesnoth.heal_unit and wesnoth.modify_ai_wml. When the corrsponding tags were 'moved' to lua we on purpose did not put them i the wiki so that we could still change/remove them later without thinking about compability.

File: src/scripting/game_lua_kernel.cpp (hunk 1 above)
Comment (CelticMinstrel) [inline src/scripting/game_lua_kernel.cpp]: It's a fourth argument, but a third _boolean_ argument since the first argument is a location.

**Comment by CelticMinstrel** [comment]
Wow, apparently I forgot to commit the actual new implementation of [animate_unit]...

**Comment by CelticMinstrel** [comment]
I think that addresses everything @gfgtdf commented on... now to move on to other tags! (And maybe also test animate_unit, I guess.)

Currently on the todo list are [kill], [shroud], and [scroll]. Subject to change at any time though.

File: src/scripting/game_lua_kernel.cpp [hunk 2]
```diff
@@ -2784,173 +2849,74 @@ int game_lua_kernel::intf_match_side(lua_State *L)
...
 
-	side_filter ssf(cfg, &game_state_);
-	std::vector<int> sides = ssf.get_teams();
-	for (const int &side_num : sides)
-	{
-		ai::manager::modify_active_ai_for_side(side_num,cfg.get_parsed_config());
+	if(flag.empty() && color.empty()) {
+		return 0;
+	}
+	team& side = teams()[team_i];
```
Comment (gfgtdf) [inline src/scripting/game_lua_kernel.cpp]: This looks like it misses need a check wheterh team_i is actuall in range of teams().

Also i think this functions name is missleading in 2 ways:
1) All the other functions use teh tern 'side' (match_side, get_sides...)
2) It doesnt chage an 'id' (side= or save_id=) of that side at all.

Comment (AI0867) [inline src/scripting/game_lua_kernel.cpp]: We typically call that 'name' or something like that.

Comment (gfgtdf) [inline src/scripting/game_lua_kernel.cpp]: If it just sets flag and color it name ir wesnoth.set_flag_and_color(side, flag, color)

Comment (CelticMinstrel) [inline src/scripting/game_lua_kernel.cpp]: @AI0867 It's not "name", it's colour and flag style.

@gfgtdf I dislike a "set_x_and_y" style of naming, but I suppose I could go with that if there are no other ideas. It'd be nice if we could come up with a single word that encompasses both the flag and the colour.

File: data/lua/wml-tags.lua [hunk 3]
```diff
@@ -688,11 +688,36 @@ function wml_actions.store_side(cfg)
...
 
--- This is the port of the old [modify_ai] into lua. It is different from wesnoth.modify_ai in that it uses a standard side filter.
--- I don't know why these functions were made to behave differently, but this seems to be the more powerful and useful one according
--- to mattsc's comments
 function wml_actions.modify_ai(cfg)
-	wesnoth.modify_ai_wml(cfg)
+	local sides = utils.get_sides(cfg)
+	local component
+	if cfg.action == "add" or cfg.action == "change" then
+		local start = string.find(cfg.path, "[a-z]+%[[a-z0-9_*]+%]$")
```
Comment (mattsc) [inline data/lua/wml-tags.lua]: This does not work in all cases.  It does not match 'candidate_action' and it does not work with empty closing []. It should be
`string.find(cfg.path, '[a-z_]+%[[a-z0-9_*]*%]$')`
I think (note: two changes).

File: src/scripting/game_lua_kernel.cpp [hunk 4]
```diff
@@ -2858,6 +2858,9 @@ int game_lua_kernel::intf_set_side_id(lua_State *L)
 	if(flag.empty() && color.empty()) {
 		return 0;
 	}
+	if(team_i < 0 || team_i >= teams().size()) {
+		return luaL_error(L, "set_side_id: side number %d out of range", team_i);
```
Comment (mattsc) [inline src/scripting/game_lua_kernel.cpp]: This produces a type mismatch warning, which causes at least the Xcode build to fail.

**Comment by Vultraz** [comment]
Please do finish this at some point.

**Pull request merged**